sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -113,31 +113,107 @@ object ScalaJSPluginInternal {
       if ((skip in taskKey).value)
         Def.task((artifactPath in taskKey).value)
       else Def.task {
-        val cp = scalaJSPreLinkClasspath.value
+        val s = (streams in taskKey).value
+        val deps = resolvedJSDependencies.value
         val output = (artifactPath in taskKey).value
-        val taskCache = WritableFileVirtualTextFile(
-            streams.value.cacheDirectory / cacheName)
 
-        IO.createDirectory(output.getParentFile)
+        val realFiles = deps.get(sourceFiles).get
+        val resolvedDeps = deps.data
 
-        val outFile = AtomicWritableFileVirtualJSFile(output)
-        CacheUtils.cached(cp.version, outFile, Some(taskCache)) {
-          toolsIO.concatFiles(outFile, cp.jsLibs.map(getLib))
-        }
+        FileFunction.cached(s.cacheDirectory, FilesInfo.lastModified,
+            FilesInfo.exists) { _ => // We don't need the files
+
+          IO.createDirectory(output.getParentFile)
+
+          val outFile = AtomicWritableFileVirtualJSFile(output)
+          toolsIO.concatFiles(outFile, resolvedDeps.map(getLib))
+
+          Set(output)
+        } (realFiles.toSet)
 
         output
       }
     }
   }
 
-  private def scalaJSOptimizerSetting(key: TaskKey[_]): Setting[_] = (
-      scalaJSOptimizer in key := {
+  /** Settings for the production key (e.g. fastOptJS) of a given stage */
+  private def scalaJSStageSettings(stage: Stage, key: TaskKey[File]): Seq[Setting[_]] = Seq(
+      scalaJSLinker in key := {
+        val opts = (scalaJSOptimizerOptions in key).value
+
         val semantics = (scalaJSSemantics in key).value
         val outputMode = (scalaJSOutputMode in key).value
-        if ((scalaJSOptimizerOptions in key).value.parallel)
-          new ScalaJSOptimizer(semantics, outputMode, ParIncOptimizer.factory)
-        else
-          new ScalaJSOptimizer(semantics, outputMode, IncOptimizer.factory)
+        val withSourceMap = (emitSourceMaps in key).value
+
+        val relSourceMapBase = {
+          if ((relativeSourceMaps in key).value)
+            Some((artifactPath in key).value.getParentFile.toURI())
+          else
+            None
+        }
+
+        val frontendConfig = LinkerFrontend.Config()
+          .withBypassLinkingErrors(opts.bypassLinkingErrors)
+          .withCheckIR(opts.checkScalaJSIR)
+          .withDisableOptimizer(opts.disableOptimizer)
+
+        val backendConfig = LinkerBackend.Config()
+          .withRelativizeSourceMapBase(relSourceMapBase)
+          .withCustomOutputWrapper(scalaJSOutputWrapper.value)
+          .withPrettyPrint(opts.prettyPrintFullOptJS)
```

Comment:
So, are leaving this be for the moment? Or should we expose keys for the front-end and back-end `Config`s in addition to `scalaJSOptimizerOptions`? I guess we can leave it as it is until we get an actual use case. In the meantime, a user can copy-paste this setting anyway, if required.